### ISSUE TITLE
Select values lost in snapshots

### ISSUE THREAD

**Issue by callumacrae** [cypress-io/cypress#4443]
### Current behavior:

When viewing a snapshot the values of `<select>` elements are lost. `<input>` element preserve their value fine, however.

This is minor, but it's happening in my percy tests too and I guess it's caused by the same thing.

### Desired behavior:

`<select>` elements should appear in snapshots as they appeared at the time.

### Steps to reproduce: (app code and test code)

Repo and steps to replicate in readme: https://github.com/callumacrae/cypress-select-value-issue

### Versions

Cypress 3.3.1

**Comment by jennifer-shehane** [comment]
Thanks for opening an issue. This is a known issue, although I am not sure there has been an issue opened for it previously - so we'll leave this issue open to track this.

**Comment by x-yuri** [comment]
A cursory search doesn't reveal similar issues. And actually, the issue is minor only when you know about it. I thought there was something with my tests, took me a while to confirm that everything's fine. Here's my [report][1] just in case :)

[1]: https://github.com/x-yuri/issues/tree/cypress-snapshots

**Comment by x-yuri** [comment]
Okay, I see, the issue is with cloning the DOM [here][1]. It all comes down to the fact that `Node.cloneNode()` doesn't clone `select`'s properties. It's so in the [spec][2] (won't fix), in [`chrome`][3] and [`ff`][5] (won't fix), and [`jquery`][4] (won't fix). The last hope is `cypress` :)

Possible workaround:

```
$origBody = $$('body')
$body = $origBody.clone()
$selects = $body.find('select')
$origBody.find('select').each(
    (i, os) => $selects[i].value = os.value
)
```

A fiddle: https://jsfiddle.net/xyuri/ktf1bqey/22/

[1]: https://github.com/cypress-io/cypress/blob/v3.4.0/packages/driver/src/cy/snapshots.coffee#L121
[2]: https://github.com/whatwg/dom/issues/644
[3]: https://bugs.chromium.org/p/chromium/issues/detail?id=838853
[4]: https://bugs.jquery.com/ticket/1294
[5]: https://bugzilla.mozilla.org/show_bug.cgi?id=1458507

@jennifer-shehane Could this make it resolved faster?

**Comment by fr0** [comment]
I disagree that this is a minor issue; being able to see the state of the select options in your app is valuable.

**Comment by callumacrae** [comment]
> can lead to wasted time trying to figure out why the selected values are wrong (which is what happened to me).

That's what happened to me, too!

**Comment by matthieu-foucault** [comment]
Running into this more than a year later. This is an issue when using visual testing solutions such as Percy or Happo. Is it also a wontfix on cypress's end?

**Comment by jennifer-shehane** [comment]
@matthieu-foucault This is not a 'wontfix', it is labeled as 'ready for work' and an open issue. It is just not prioritized currently.

**Comment by fringd** [comment]
Any clues as to where a brave soul would start fixing this?

**Comment by x-yuri** [comment]
@fringd Doesn't [this][a] answer your question?

[a]: https://github.com/cypress-io/cypress/issues/4443#issuecomment-510268204

**Comment by crfrolik** [comment]
Please do not close this. It is still a bug that has not been fixed yet.

**Comment by nagash77** [comment]
@crfrolik thank you for confirming this is still an issue.  I will update the ticket accordingly.

**Comment by iomedico-beyer** [comment]
I have huge problems to debug my code because of this issue. And sadly I do not understand [the idea of x-yuri](https://github.com/cypress-io/cypress/issues/4443#issuecomment-510268204). Can this somehow be translated into a custom select command?

And a related question: if (while tested in a loop) `cy.get("foo").select("bar"); cy.get("foo option[var='bar']").should("be.selected")` sometimes fails but `cy.get("foo").select("bar").invoke("val").should("eq", "bar")` always passes, can I assume the select went always well? I do not understand what happens, because I cannot visually verify the snapshots. But I guess the answer is no as the dom may have changed between the two `get` commands. In fact I think it does, because calling the second `get` in `within` fails with “The subject is no longer attached to the DOM, and Cypress cannot requery the page after commands such as cy.within().”

**Comment by x-yuri** [comment]
> And sadly I do not understand the idea of x-yuri

My idea is to change the `cypress` code to make it work. Find [this line][a] in the installed `cypress` and replace with the code from [here][b]. If you're using `cypress-3.4.0` it should make it work. For newer versions you might need to do some extra steps.

> Can this somehow be translated into a custom select command?

Unlikely, but not 100% sure.

[a]: https://github.com/cypress-io/cypress/blob/v3.4.0/packages/driver/src/cy/snapshots.coffee#L121
[b]: https://github.com/cypress-io/cypress/issues/4443#issuecomment-510268204
